From c5f51f6b0d2d4cd8b2ec466c8c7628bc35ff88db Mon Sep 17 00:00:00 2001
From: Ricardo Cerqueira <cyanogenmod@cerqueira.org>
Date: Wed, 28 May 2014 17:12:24 +0100
Subject: [PATCH 1/2] wcnss_qmi: Generate a fixed random mac address if the NV
 doesn't provide one

If the NV is corrupted or blanked out, QMI won't return a MAC address
and prima will fallback to a fixed one. If the board defines a
TARGET_WCNSS_MAC_PREFIX value (2 to 4 hex values), use that prefix
to generate a random (and persistent) hwaddr instead.

For "TARGET_WCNSS_MAC_PREFIX := 0a0b0c0d", the last 2 bytes will be
pseudo-randomized, and the final address will be in the
"0a:0b:0c:0d:XX:YY" format.

Change-Id: Ie38cb5d3724223b2ec1ae45e6a9a4f2285229731
---
 wcnss-service/Android.mk         |  5 +++
 wcnss-service/wcnss_qmi_client.c | 57 ++++++++++++++++++++++++++++++++
 2 files changed, 62 insertions(+)

diff --git a/wcnss-service/Android.mk b/wcnss-service/Android.mk
index 1239015..f644110 100644
--- a/wcnss-service/Android.mk
+++ b/wcnss-service/Android.mk
@@ -56,6 +56,11 @@ LOCAL_SHARED_LIBRARIES += libqmiservices libqmi_cci
 LOCAL_C_INCLUDES += $(TARGET_OUT_HEADERS)/libmdmdetect/inc
 LOCAL_SHARED_LIBRARIES += libmdmdetect
 LOCAL_HEADER_LIBRARIES += libril-qc-qmi-services-headers
+
+ifneq ($(TARGET_WCNSS_MAC_PREFIX),)
+    LOCAL_CFLAGS += -DWCNSS_INVALID_MAC_PREFIX=\"$(TARGET_WCNSS_MAC_PREFIX)\"
+endif
+
 LOCAL_CFLAGS += -DWCNSS_QMI
 LOCAL_SRC_FILES += wcnss_qmi_client.c
 
diff --git a/wcnss-service/wcnss_qmi_client.c b/wcnss-service/wcnss_qmi_client.c
index 3e1fb0e..5a9e0db 100644
--- a/wcnss-service/wcnss_qmi_client.c
+++ b/wcnss-service/wcnss_qmi_client.c
@@ -29,6 +29,8 @@ ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #ifdef WCNSS_QMI
 #define LOG_TAG "wcnss_qmi"
 #include <cutils/log.h>
+#include <stdlib.h>
+#include <sys/stat.h>
 #include "wcnss_qmi_client.h"
 #include "qmi_client.h"
 #include "device_management_service_v01.h"
@@ -107,8 +109,63 @@ int wcnss_qmi_get_wlan_address(unsigned char *pBdAddr)
 		ALOGE("%s: Succesfully Read WLAN MAC Address", __func__);
 		return SUCCESS;
 	} else {
+#ifdef WCNSS_INVALID_MAC_PREFIX
+#ifndef WCNSS_GENMAC_FILE
+#define WCNSS_GENMAC_FILE "/persist/.genmac"
+#endif
+		int i = 0;
+		struct stat statbuf;
+		FILE *genmac;
+		int macbytes[6] = { 0, };
+		// Limit the prefix to 4 bytes, we want at least 2 to be random
+		int prefixlen = strnlen(WCNSS_INVALID_MAC_PREFIX,8)/2;
+
+		// Misconfigured device source...?
+		if (prefixlen < 2) {
+			return FAILED;
+		}
+
+		// Use a previously stored value if it exists
+		if (!stat(WCNSS_GENMAC_FILE, &statbuf)) {
+			genmac = fopen(WCNSS_GENMAC_FILE,"r");
+			if (fscanf(genmac, "%c%c%c%c%c%c", &pBdAddr[0],
+				     &pBdAddr[1], &pBdAddr[2], &pBdAddr[3],
+				     &pBdAddr[4], &pBdAddr[5]) == 6) {
+				fclose(genmac);
+				ALOGE("%s: Succesfully Read local WLAN MAC Address", __func__);
+				return SUCCESS;
+			}
+			fclose(genmac);
+		}
+
+		sscanf(WCNSS_INVALID_MAC_PREFIX, "%2x%2x%2x%2x",
+				&macbytes[0], &macbytes[1],
+				&macbytes[2], &macbytes[3]);
+
+		// We don't need strong randomness, and if the NV is corrupted
+		// any hardware values are suspect, so just seed it with the
+		// current time
+		srand(time(NULL));
+
+		for (i = prefixlen; i<6; i++) {
+			macbytes[i] = rand() % 255;
+		}
+		// Invert them
+		for (i = 0; i < 6; i++) {
+			pBdAddr[i] = macbytes[5-i];
+		}
+
+		// Store for reuse
+		genmac = fopen(WCNSS_GENMAC_FILE,"w");
+		fwrite(pBdAddr, 1, 6, genmac);
+		fclose(genmac);
+
+		ALOGE("%s: Failed to Read WLAN MAC Address, successfully randomized one", __func__);
+		return SUCCESS;
+#else
 		ALOGE("%s: Failed to Read WLAN MAC Address", __func__);
 		return FAILED;
+#endif
 	}
 }
 
-- 
2.19.1


From ca432f1a9949724622c8e8093cb928b1ecc18978 Mon Sep 17 00:00:00 2001
From: Ethan Chen <intervigil@gmail.com>
Date: Sun, 24 Sep 2017 18:29:55 -0700
Subject: [PATCH 2/2] wcnss_service: Read serial number from custom property

* ro.serialno and ro.boot.serialno are labeled as serialno_prop, which
  is restricted in the domains which are allowed to read it.
* Add a duplicate serial number property for wcnss_service to read.

Change-Id: I0b0cd68c399d2e711788c8c1f91c81e7db2f06e9
---
 wcnss-service/wcnss_service.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/wcnss-service/wcnss_service.c b/wcnss-service/wcnss_service.c
index f064fa1..eafa81b 100644
--- a/wcnss-service/wcnss_service.c
+++ b/wcnss-service/wcnss_service.c
@@ -417,7 +417,7 @@ void setup_wcnss_parameters(int *cal)
 		return;
 	}
 
-	rc = property_get("ro.serialno", serial, "");
+	rc = property_get("ro.wcnss.serialno", serial, "");
 	if (rc) {
 		serial_num = convert_string_to_hex(serial);
 		ALOGE("Serial Number is  %x", serial_num);
-- 
2.19.1

